tests/test.ml

```diff
@@ -16,4 +16,55 @@ let () =
   assert (getOption (index 12) [0;1;2;3;4] = None);
   assert (getOption (index ~-1) [0;1;2;3;4] = None);
   assert (set (index 3) '!' "abcdef" = "abc!ef");
-  assert (getOption (index 3) "abcdef" = Some 'd')
+  assert ("abcdef" ^? index 3 = Some 'd');
+  assert (('a', 'b') ^? index 0 = Some 'a');
+  assert (('a', 'b') ^? index 3 = None);
+  assert (set (index 0) 'c' ('a', 'b') = ('c', 'b'));
+  assert (set (index 3) 'c' ('a', 'b') = ('a', 'b'));
+  assert (set (mapped {Imp.Control.List}) 4 [1;2;3] = [4;4;4]);
+  assert (("hello", 5) |> T2._2 @~ ((+) 1) = ("hello", 6));
+  assert (("hi", "five") |> T2._2 @. 5 = ("hi", 5));
+  assert (("hi", 5) ^. T2._1 = "hi");
+  assert (("hi", 5) |> T2._1 @? true = (Some true, 5));
+  assert (123 ^? empty = None);
+  assert ("five" |> equality @. 5 = 5);
+  assert ([1;2;3] |> head @. 0 = [0;2;3]);
+  assert ([1;2;3] |> tail @. 0 = [1;0;0])
+
+let () =
+  let open Imp.Any in
+  let open Lens in
+  let module IntOrd = struct
+    type t = int
+    let compare (x: int) y = compare x y
+  end in
+  let implicit module IntMap = Map.Make (IntOrd) in
+  assert (IntMap.empty ^? index 3 = (None: char option));
+  let myMap = IntMap.(empty |> add 1 "one" |> add 2 "two") in
+  assert (myMap ^. at 1 = Some "one");
+  assert (myMap ^. at 3 = (None: string option));
+  assert (myMap |> at 1 @? "one" = myMap);
+  let myMap' = myMap |> IntMap.add 3 "three" in
+  assert (myMap |> at 3 @? "three" = myMap');
+  assert (myMap' |> at 3 @. None = myMap);
+  (* without this, we get an unused open Imp.Any;
+     and changing that to open implicit Imp.Any gives a type error for some reason *)
+  Any_Int.__any__
+
+let () =
+  let open Imp.Any in
+  let open Imp.Control in
+  let open Imp.Transformers in
+  let open Lens in
+  let module S = State {Any_Option {Any_Int}} in
+  let (>>) a b = bind {S} a (fun _ -> b) in
+  let t = traversed {Imp.Control.Option} in
```

Comment:
It's annoying that you can't pass something of type `{M: T} -> t` where something of type `t` is expected and have it infer `M`, even though you can workaround it by changing it to `{M: T} -> unit -> t`, and calling it with `()`, so it now just has type `t`.

Assuming suitable modules and module types:
```ocaml
let x {M: T} = M.x
let y (x: int) = x + 1
y x (* doesn't work *)


let x {M: T} () = M.x
let y (x: int) = x + 1
y (x ()) (* works *)
```

Reply:
Yes, the need to pass `()` is sometimes less convenient.  The reason it's designed that way is that passing an implicit argument can cause some effects to occur (consider: `let f {M} = perform_effects (); v`), so we chose to allow implicit arguments to be instantiated only in explicit function calls, not in expressions that look like values.